src/urls/package-url-handler.ts

```diff
@@ -0,0 +1,149 @@
+/**
+ * @license
+ * Copyright (c) 2017 The Polymer Project Authors. All rights reserved.
+ * This code may only be used under the BSD style license found at
+ * http://polymer.github.io/LICENSE.txt
+ * The complete set of authors may be found at
+ * http://polymer.github.io/AUTHORS.txt
+ * The complete set of contributors may be found at
+ * http://polymer.github.io/CONTRIBUTORS.txt
+ * Code distributed by Google as part of the polymer project is also
+ * subject to an additional IP rights grant found at
+ * http://polymer.github.io/PATENTS.txt
+ */
+
+import {lookupDependencyMapping} from '../manifest-converter';
+
+import {ConvertedDocumentFilePath, ConvertedDocumentUrl, OriginalDocumentUrl, PackageType} from './types';
+import {UrlHandlerInterface} from './url-handler-interface';
+import {getRelativeUrl} from './util';
+
+const isInBowerComponentsRegex = /(\b|\/|\\)(bower_components)(\/|\\)/;
+const isInNodeModulesRegex = /(\b|\/|\\)(node_modules)(\/|\\)/;
+
+/**
+ * Handle URLs in a single "package-based" layout. This converter should be used
+ * to convert a single package, where all existing Bower dependencies are
+ * installed in a "bower_components/" sub-directory inside the main package
+ * directory.
+ */
+export class PackageUrlHandler implements UrlHandlerInterface {
+  readonly packageName: string;
+  readonly packageType: PackageType;
+
+  /**
+   * Helper function to check if a file URL is internal to the main package
+   * being converted (vs. a dependency).
+   */
+  static isUrlInternalToPackage(url: ConvertedDocumentUrl|OriginalDocumentUrl|
+                                ConvertedDocumentFilePath) {
+    return !isInBowerComponentsRegex.test(url) &&
+        !isInNodeModulesRegex.test(url);
+  }
+
+  constructor(packageName: string, packageType?: PackageType) {
+    this.packageName = packageName;
+    this.packageType = packageType || 'element';
+  }
+
+  /**
+   * Get the converted NPM name for a package, given the original document URL
+   * of any file that lives in that package.
+   */
+  getPackageNameForUrl(url: OriginalDocumentUrl) {
+    if (PackageUrlHandler.isUrlInternalToPackage(url)) {
+      return this.packageName;
+    } else {
+      // For a a single package layout, the Bower package name is included in
+      // the URL. ie: bower_components/PACKAGE_NAME/...
+      const basePackageName = url.split('/')[1];
```

Comment:
I feel like TypeScript should have yelled at you for not ensuring that `basePackageName` is defined (i.e. if url didn't contain a `/`) since `lookupDependencyMapping` demands a string for its first argument.

We must have less strict settings or something on tsconfig/tslint for this project?  I'd have expected `url.split('/')[1]!` or something.